### PR TITLE
move source generation into separate module

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
refactoring

**Did you add tests for your changes?**
existing tests

**If relevant, link to documentation update:**
N/A

**Summary**
move source generation into separate module
allow to pass a Generator to NormalModule
change NormalModule constructor to options object

**Does this PR introduce a breaking change?**
yes

**Other information**
cc @michael-ciniawsky

### CODE REVIEW

File: lib/JavascriptGenerator.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+const RawSource = require("webpack-sources").RawSource;
+const ReplaceSource = require("webpack-sources").ReplaceSource;
+
+// TODO: clean up this file
+// replace with newer constructs
+
+// TODO: remove DependencyVariables and replace them with something better
+
+class JavascriptGenerator {
```
Comment (ooflorent) [inline lib/JavascriptGenerator.js]: JavaScript is written with a capital S. Every time I see it written this way in webpack I need to breath deeply 😅

File: lib/JavascriptGenerator.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		return "this";
+	}
+
+	/*
+	 * creates the end part of a IIFE around the module to inject a variable name
+	 * (function(...){
+	 * }.call(...))   <- this part
+	 */
+	variableInjectionFunctionWrapperEndCode(module, varExpressions, block) {
+		const firstParam = this.contextArgument(module, block);
```
Comment (ooflorent) [inline lib/JavascriptGenerator.js]: `contextParam`  would be easier to understand what's going on.

**Comment by TheLarkInn** [comment]
The API for this is a bit verbose, however I think it does serve the needs that we have for now, and lays the foundation for x-as-module type that's non-js-like.

File: lib/JavascriptGenerator.js (hunk 1 above)
Comment (TheLarkInn) [reply lib/JavascriptGenerator.js]: PascalCase ignores propernow capitalization I believe. Since this is only a two word name.

Comment (ooflorent) [reply lib/JavascriptGenerator.js]: So, can we talk about WebAssembly? 😉

Comment (TheLarkInn) [reply lib/JavascriptGenerator.js]: Touche. Haha well we should just update them all in one PR, not this one though. 😂

**Pull request merged**